text_generation/causal_lm/cpp/continuous_batching/library/src/sampler.hpp

```diff
@@ -198,7 +201,90 @@ class GroupBeamSearcher {
     }
 };
 
+using LogitWithIdx = std::pair<float, size_t>;
+using ProbabilityWithIdx = std::pair<float, size_t>;
+
+class IProbabilityFilter {
+public:
+    virtual std::vector<LogitWithIdx> filter(const std::vector<LogitWithIdx>& input_logits) = 0;
+};
+
+
+class TopPFilter: public IProbabilityFilter {
+public:
+    TopPFilter(double top_p) : m_top_p(top_p) {
+        OPENVINO_ASSERT(top_p > 0.0f && top_p <= 1.0f, "top_p must be in the interval (0, 1]");
+    }
+
+    std::vector<ProbabilityWithIdx> filter(const std::vector<ProbabilityWithIdx>& input_probs) override {
+        std::vector<ProbabilityWithIdx> tmp(input_probs);
+        std::sort(tmp.begin(), tmp.end(), [](const ProbabilityWithIdx& lhs, const ProbabilityWithIdx& rhs) {return lhs.first > rhs.first; });
+        float probability_sum = 0.0f;
+        size_t nucleus_size = 0;
+        for (const auto& probability : tmp) {
+            probability_sum += probability.first;
+            nucleus_size += 1;
+            if (probability_sum > m_top_p) break;
+        }
+        return std::vector<ProbabilityWithIdx>(tmp.begin(), tmp.begin() + nucleus_size);
+    }
+
+private:
+    double m_top_p;
+};
+
+class TopKFilter: public IProbabilityFilter {
+public:
+    TopKFilter(size_t top_k) : m_top_k(top_k) {}
+
+    std::vector<ProbabilityWithIdx> filter(const std::vector<ProbabilityWithIdx>& input_probs) override {
+        std::vector<ProbabilityWithIdx> tmp(input_probs);
+        std::sort(tmp.begin(), tmp.end(), [](const ProbabilityWithIdx& lhs, const ProbabilityWithIdx& rhs) {return lhs.first > rhs.first; });
+        size_t top_k = input_probs.size() >= m_top_k ? m_top_k : input_probs.size();
+        return std::vector<ProbabilityWithIdx>(tmp.begin(), tmp.begin() + top_k);
+    }
+
+private:
+    size_t m_top_k;
+};
+
+class TemperatureLogitTransform {
```

Comment:
HF implementation also applies repetition_penalty for sampling decoding:
https://github.com/huggingface/transformers/blob/main/src/transformers/generation/logits_process.py#L296